README.md
# IKAN ALM Client Rest API plugin

[![Jenkins Plugin](https://img.shields.io/jenkins/plugin/v/ikan-almclient-downloader.svg)](https://plugins.jenkins.io/ikan-almclient-downloader) [![GitHub release](https://img.shields.io/github/v/release/jenkinsci/ikan-almclient-downloader.svg?label=release)](https://github.com/jenkinsci/ikan-almclient-downloader-plugin/releases) [![Jenkins Plugin Installs](https://img.shields.io/jenkins/plugin/i/ikan-almclient-downloader.svg?color=blue)](https://plugins.jenkins.io/ikan-almclient-downloader)

## Overview

The ALM Client RestAPI plugin allows Builder’s Users to execute IKAN ALM main Rest API's, such as:
Create Level Request
Project Stream 
Package
Approve or Reject a Level Request
List Level Request Issues.
This will allow users to run IKAN ALM Build and Deploy requests and other tasks with their Builder Software. 

## Prerequisites

The following are required to use this plugin:

* Builder Software (Jenkins/CloudBees/Hudson)
* [IKAN ALM Software](https://www.ikanalm.com "IKAN ALM") 
* IKAN ALM Base Plugin
* IKAN ALM Client RestAPI Plugin


## Installing in a Builder Instance

First method from IKAN site:

- Ask the last JENKINS-solution_ALMCLIENT archive to [IKAN site](https://www.ikan.be "IKAN") using our *Contact us* menu.
- Unzip the received JENKINS-solution_ALMCLIENT-x.y.z.zip (tar.gz) archive in the Builder server.
- Install the ALM Base Plugin (almbase-jenkins-plugin-x.y.z-SNAPSHOT.hpi)  
and ALM Client RestAPI Plugin (ikan-almclient-plugin-x.y.z-SNAPSHOT.hpi) according to the Builder instructions for installing plugins. Go to Manage Plugins menu  
and Advanced settings for selecting each hpi file in the Deploy plugin part of the page.

Second method from ‘Available plugins’:

- Install the ALM Base Plugin using the ‘Available plugins’ in your Builder Software, first.
- Install the ALM Client RestAPI Plugin using the ‘Available plugins’ in your Builder Software, next.
- Ask the last Pipeline models to [IKAN site](https://www.ikan.be "IKAN") using our *Contact us* menu or by our ALM Support <support@ikanalm.com>. 

## Configuring Administration

### Configuration System

Set the variables as:

- Environment variables like:

---
	Name	almBaseUrl
	Value	http://ikan000:8080/alm

---
	Name	almSystem_location
	Value	C:\Jenkins\almPluginBase
---

The **almSystem_location** parameter must have the location where the JENKINS-solution_ALMCLIENT archive has been unzipped with the *almPluginsBase* folder. If you don’t use this archive, you create this folder somewhere in your Server. Next you create a *Pipelines* folder for storing provided pipeline models.

### Manage Plugins

- Activate if not: 	
    - ‘Build Pipeline Plugin’
	- ‘Build Timeout’
	- ‘Pipeline Utility Steps’	(if not exist, download it, and add it for your Jenkins version)
	- ‘Workspace Cleanup Plugin’
	- ‘Artifactory Plugin’	(Optional)

- Install ALM Plugins with the “Advanced” sheet (first way) or “Available” sheet (second way):
	- ALM Base Plugin ‘ikan-almbase-plugin’
	- ALM Client RestAPI Plugin ‘ikan-almclient-plugin’

### Global Tool Configuration

Add if not: 	‘JDK’ Installation

	Name		jdk-11.0.7
	JAVA-HOME	C:\Java\jdk-11.0.7

Add if not: 	‘ANT’ Installation

	Name		ANT-1.10.10
	Install automatically
	Install from Apache
	Version		1.10.10

## Executing IKAN ALM Client Rest API activities

Create a ‘Pipeline’ project and use the corresponding Pipeline script model provided in the archive.

	- almPipeline_almclientApproval.groovy
	- almPipeline_almclientLevelRequest.groovy
	- almPipeline_almclientLevelRequestIssues.groovy
	- almPipeline_almclientPackage.groovy
	- almPipeline_almclientProjectStream.groovy

These pipelines are prepared to manage all parameters used by the task to run.
First, you set correct values for:

```
	def antInstallation = 'ANT-1.10.10'
	def jdkInstallation = 'jdk-11.0.7'
```

Next, each parameter is defined with or without ‘//’ (groovy comment).  
When a parameter is commented, it must be defined in the Parameters of the Project with valid structure.  
(Choice, Password, String, etc.). Other parameters are "hard coded" in the pipeline for the steps.

## Using Pipeline Syntax to Generate Pipeline Script

1. Do one of the following:
	* When working with an existing Pipeline job, click the *Pipeline Syntax* link in the left panel. The *Snippet Generator* appears.
	* When configuring a Pipeline job, click the *Pipeline Syntax* link at the bottom of the Pipeline configuration section. The *Snippet Generator* appears.
2. From the *Sample Step* list, select one *almclientXxx* task.
3. From the *Parameters* list, select or provide the correct value for this task.
4. Click *Generate Pipeline Script* button. The Groovy script to invoke the almclientXxx script appears.  
The script can be added to the Pipeline section when configuring a Pipeline job.  
A sample script is shown below:

![Generate Pipeline Script](./images/GeneratePipelineScript.PNG)

5.  Copy the script text and put it into your Job pipeline part. Next see the Pipeline Requipement paragraph for completing the Job pipeline lines of the step with this example.

## Pipeline Requirement

To use a generated pipeline, your almBaseURL, almRestUserid and almRestPassword must be configured properly to run it with other useful parameters. Define them in the configure system screen or in the Project parameters.
Next the task must be surround with these lines:

```
            steps {
                script {
                    propertyFileALM = "${sourceProjectDir}/almclientLevelRequest.properties"
                }
                dir("${sourceProjectDir}") { writeFile file:'almclientLevelRequest.properties', text:'' }
                echo pwd()
       	        echo "JAVA_HOME=${JAVA_HOME}"
	        echo "Start almclientLevelRequest()"
                withEnv(["ANT_ARGS=-q -propertyfile ${propertyFileALM}"]) {
    		    withAnt(jdk: "${jdkInstallation}") {
			almclientXxx (
			    ...
            )
		}
	}

```

### Pipeline LevelRequest sample

```
def antInstallation = 'ANT-1.10.10'
def jdkInstallation = 'jdk-11.0.7'

pipeline {
	// ------------------------------
	// IKAN ALM LevelRequest
	// ------------------------------
	// specify agents/nodes
    agent any
	// define scripting tools
	tools {
		ant "${antInstallation}"
		jdk "${jdkInstallation}"
	}

    environment {
        // almBaseUrl: 'http://IKAN000:8080/alm'
        // almSystem_location
        // almRestPassword= ''
        // almRestUserid= ''
        // almStartDate= ''
        // almStatus= ''
        // almVcrTag= ''
        // builderAction = 'Create'
        builderName = 'Jenkins'
        // buildNumber = '1'
        // buildPrefix = ''
        // buildSuffix = ''
        // deployEnvironments = ''
        // levelName = 'TEST'
        // levelRequestOid = ''
        // packageName = ''
        // paramfile= 'parameters.properties'
        // projectName = 'thisProject'
        // redeliver = 'false'

        sourceProjectDir = "$almSystem_location/environments/build/$projectName"
    }
    stages {
        stage('Create LevelRequest') {
            steps {
                script {
                    propertyFileALM = "${sourceProjectDir}/almclientLevelRequest.properties"
                }
                dir("${sourceProjectDir}") { writeFile file:'almclientLevelRequest.properties', text:'' }
                echo pwd()
       	        echo "JAVA_HOME=${JAVA_HOME}"
	        echo "Start almclientLevelRequest()"
                withEnv(["ANT_ARGS=-q -propertyfile ${propertyFileALM}"]) {
    		    withAnt(jdk: "${jdkInstallation}") {
                        almclientLevelRequest (
                            antInstallation: "${antInstallation}",
                            almBaseUrl: '$almBaseUrl',
                            almBuildNumber: '$buildNumber',
                            almDeploysToExecute: '$deployEnvironments',
                            almDescription: 'Build created by $builderName with Rest API',
                            almLevelRequestLevelName: '$levelName', 
                            almLevelRequestOid: '$levelRequestOid', 
                            almPackageName: '$packageName', 
                            almProjectName: '$projectName', 
                            almProjectStreamBuildPrefix: '$buildPrefix', 
                            almProjectStreamBuildSuffix: '$buildSuffix',
                            almRedeliver: '$redeliver',
                            almRestPassword: '$almRestPassword',
                            almRestUserid: '$almRestUserid',
                            almStartDate: '$almStartDate',
                            almStatus: '$almStatus',
                            almStopOnError: 'true',
                            almVcrTag: '$almVcrTag',
                            builderAction: '$builderAction',
                            builderName: '$builderName',
                            levelRequestAccess: 'lroid',
                            paramfile: '$paramfile',
                            statusMaxwaitMinutes: '30',
                            statusSuccessMandatory: 'true',
                            statusWaitingSeconds: '30'
                        )
                    }
                }
                echo "End almclientLevelRequest()"
            }
        }
    }
}

```

# Product Documentation

All IKANALM product documentation can be found here: https://docs.ikanalm.com/

IKAN ALM REST API v1 Documentation: https://docs.ikanalm.com/rest-api/5.9/RESTAPI_Overview.html


## Contacting Customer Support

If you need support services, please collect the following information before contacting IKAN’s support:  

* Your name, your e-mail, the subject will contain the name, release number of your product.
* the message will contain if necessary:
    - Installation information including installed options, whether the product uses local or network databases, whether it is installed in the default directories, whether it is a standalone or network installation, and whether it is a client or agent or server installation.
    - Environment information, such as the operating system and release on which the product is installed, memory, hardware and network specification, and the names and releases of other applications that were running when the problem occurred.
    - The location of the problem within the running application and the user actions taken before the problem occurred.
    - The exact application, licensing, or operating system error messages, if any.

You can report issues via following email: support@ikanalm.com


### Corporate Web Site

To access IKAN's site visit: https://www.ikanalm.com/ 


CHANGELOG.md

# Change Log

## Version 2.1.0

* The Builder plugin integrates with the ALM Client RestAPI Plugin which allows the IKAN ALM connection configuration to be defined in Configure System screen with Environment variables parameters.
    - almBaseUrl      (mandatory)
    - almSystem_location  (mandatory)
    - almRestUserid       (optional. Preferred in the Project)
    - almRestPassword (optional. Preferred in the Project)
* The Global Tools Configuration would contain an ANT and JDK 11 definitions which will be set in the Pipeline’s groovy scripts for running plugin step(s).
Pipeline samples are available in the [GitHub project](https://github.com/jenkinsci/ikan-almclient-plugin) Pipelines folder.
* Support for the Builder Pipeline Syntax and pipeline models are available for each task.
* Jenkins console logs produced by the plugin have been streamlined to improve readability.


src/main/java/com/ikanalm/plugins/jenkins/almclientapproval/AlmclientApprovalStep.java
/**
 * 
 */
package com.ikanalm.plugins.jenkins.almclientapproval;


import org.kohsuke.stapler.DataBoundConstructor;
import org.kohsuke.stapler.DataBoundSetter;
import org.kohsuke.stapler.QueryParameter;

import com.ikanalm.plugins.base.steps.BaseAntStep;

import hudson.Extension;
import hudson.Util;
import hudson.tasks.Ant;
import hudson.tasks.Ant.AntInstallation;
import hudson.util.FormValidation;
import hudson.util.ListBoxModel;
import jenkins.model.Jenkins;

/**
 * Class that defines the ALM ALM Client Level Request Approval Phase as a Jenkins pipeline step.
 * Example usage :
 * almclientApproval (antInstallation: "${antInstallation}",
 *      alm.baseUrl: "http://IKAN000:8080/alm",
 *      alm.levelRequest.levelName: "${levelName}",
 *      alm.levelRequest.oid: "",
 *      alm.package.name: "${packageName}",
 *      alm.project.name: "${projectName}",
 *      alm.projectStream.buildPrefix: "${buildPrefix}",
 *      alm.projectStream.buildSuffix: "${buildSuffix}",
 *      alm.rest.password: "",
 *      alm.rest.userid: "",
 *      alm.stopOnError: "true",
 *      approveLevelRequest: "true",
 *      approveReason: "",
 *      builderName: "Client",
 *      levelRequest_access: "lroid"
 * )
 * 
 * Note that this Step CANNOT be used in a Jenkins Freestyle Project !
 *
 * @author IKAN
 *
 */
public class AlmclientApprovalStep extends BaseAntStep {

	private static final String MAIN_SCRIPT = "almclientApproval.xml";
	private static final String FQ_PHASE_NAME = "com.ikanalm.plugins.jenkins.almclientapproval";

    private String antInstallation;
    private String almBaseUrl;
    private String almLevelRequestLevelName;
    private String almLevelRequestOid;
    private String almPackageName;
    private String almProjectName;
    private String almProjectStreamBuildPrefix;
    private String almProjectStreamBuildSuffix;
    private String almRestPassword;
    private String almRestUserid;
    private String almStopOnError;
    private String approveLevelRequest;
    private String approveReason;
    private String builderName;
    private String levelRequestAccess;

    @DataBoundConstructor
	public AlmclientApprovalStep(String antInstallation, String almBaseUrl, String almProjectName) {
	        this.antInstallation = antInstallation;
                this.almBaseUrl = almBaseUrl;
                this.almProjectName = almProjectName;

	}
        
	@Extension
	public static class DescriptorImpl extends BaseDescriptorImpl {
        /**
         * Display name of the Step in the Pipeline Syntax Snippet Generator
         */
		@Override 
		public String getDisplayName() {
            return "ALM Client Level Request Approval";
        }
		/**
		 * Name of the function that is used to call the Step in a Pipeline script 
		 */
		@Override
		public String getFunctionName() {
			return "almclientApproval";
		}
        /**
         * method that populates the configuration select named "antInstallation" with Ant installations
         * 
         * @return the antInstallation items
         */
        public ListBoxModel doFillAntInstallationItems() {
            ListBoxModel items = new ListBoxModel();
        	AntInstallation[] antInstallations = Jenkins.getInstance().getDescriptorByType(Ant.DescriptorImpl.class).getInstallations();
            
        	for (AntInstallation antInstallation : antInstallations) {
				items.add(antInstallation.getName());
			}
        	
            return items;
        }

        /**
         * Field Validation
         * @param value the antInstallation
         * @return the antInstallation description
         */
		public FormValidation doCheckAntInstallation(@QueryParameter String value) {
			if (Util.fixEmptyAndTrim(value) == null) {
				return FormValidation.warning("Choose Ant Installation");
			}
			return FormValidation.ok();
		}

		public FormValidation doCheckAlmBaseUrl(@QueryParameter String value) {
			if (Util.fixEmptyAndTrim(value) == null) {
				return FormValidation.warning("IKAN ALM URL Login cannot be empty");
			}
			return FormValidation.ok();
		}

		public FormValidation doCheckAlmLevelRequestOid(@QueryParameter String value) {
			if (Util.fixEmptyAndTrim(value) == null) return FormValidation.ok();
			try {
				Integer.parseInt(value);
				return FormValidation.ok();
			} catch (NumberFormatException e) {
				return FormValidation.error("LR Oid parameter used when levelRequest_access=lroid not numeric");
			}
		}

		public FormValidation doCheckAlmProjectName(@QueryParameter String value) {
			if (Util.fixEmptyAndTrim(value) == null) {
				return FormValidation.warning("ALM Project Name where the Level Request is created or updated. cannot be empty");
			}
			return FormValidation.ok();
		}

	}

	/* (non-Javadoc)
	 * @see com.ikanalm.plugins.jenkins.base.BaseAntStep#getFQStepName()
	 */
	@Override
	public String getFQStepName() {
		return FQ_PHASE_NAME;
	}

	/* (non-Javadoc)
	 * @see com.ikanalm.plugins.jenkins.base.BaseAntStep#getMainScript()
	 */
	@Override
	public String getMainScript() {
		return MAIN_SCRIPT;
	}
	/**
	 * @return the antInstallation
	 */
	public String getAntInstallation() {
		return antInstallation;
	}

	/**
	 * @param antInstallation the antInstallation to set
	 */
	public void setAntInstallation(String antInstallation) {
		this.antInstallation = antInstallation;
	}

	/**
	 * @return the almBaseUrl
	 */
	public String getAlmBaseUrl() {
		return almBaseUrl;
	}
	/**
	 * @param almBaseUrl the almBaseUrl to set
	 */
	@DataBoundSetter
	public void setAlmBaseUrl(String almBaseUrl) {
		this.almBaseUrl = almBaseUrl;
	}

	/**
	 * @return the almLevelRequestLevelName
	 */
	public String getAlmLevelRequestLevelName() {
		return almLevelRequestLevelName;
	}
	/**
	 * @param almLevelRequestLevelName the almLevelRequestLevelName to set
	 */
	@DataBoundSetter
	public void setAlmLevelRequestLevelName(String almLevelRequestLevelName) {
		this.almLevelRequestLevelName = almLevelRequestLevelName;
	}

	/**
	 * @return the almLevelRequestOid
	 */
	public String getAlmLevelRequestOid() {
		return almLevelRequestOid;
	}
	/**
	 * @param almLevelRequestOid the almLevelRequestOid to set
	 */
	@DataBoundSetter
	public void setAlmLevelRequestOid(String almLevelRequestOid) {
		this.almLevelRequestOid = almLevelRequestOid;
	}

	/**
	 * @return the almPackageName
	 */
	public String getAlmPackageName() {
		return almPackageName;
	}
	/**
	 * @param almPackageName the almPackageName to set
	 */
	@DataBoundSetter
	public void setAlmPackageName(String almPackageName) {
		this.almPackageName = almPackageName;
	}

	/**
	 * @return the almProjectName
	 */
	public String getAlmProjectName() {
		return almProjectName;
	}
	/**
	 * @param almProjectName the almProjectName to set
	 */
	@DataBoundSetter
	public void setAlmProjectName(String almProjectName) {
		this.almProjectName = almProjectName;
	}

	/**
	 * @return the almProjectStreamBuildPrefix
	 */
	public String getAlmProjectStreamBuildPrefix() {
		return almProjectStreamBuildPrefix;
	}
	/**
	 * @param almProjectStreamBuildPrefix the almProjectStreamBuildPrefix to set
	 */
	@DataBoundSetter
	public void setAlmProjectStreamBuildPrefix(String almProjectStreamBuildPrefix) {
		this.almProjectStreamBuildPrefix = almProjectStreamBuildPrefix;
	}

	/**
	 * @return the almProjectStreamBuildSuffix
	 */
	public String getAlmProjectStreamBuildSuffix() {
		return almProjectStreamBuildSuffix;
	}
	/**
	 * @param almProjectStreamBuildSuffix the almProjectStreamBuildSuffix to set
	 */
	@DataBoundSetter
	public void setAlmProjectStreamBuildSuffix(String almProjectStreamBuildSuffix) {
		this.almProjectStreamBuildSuffix = almProjectStreamBuildSuffix;
	}

	/**
	 * @return the almRestPassword
	 */
	public String getAlmRestPassword() {
		return almRestPassword;
	}
	/**
	 * @param almRestPassword the almRestPassword to set
	 */
	@DataBoundSetter
	public void setAlmRestPassword(String almRestPassword) {
		this.almRestPassword = almRestPassword;
	}

	/**
	 * @return the almRestUserid
	 */
	public String getAlmRestUserid() {
		return almRestUserid;
	}
	/**
	 * @param almRestUserid the almRestUserid to set
	 */
	@DataBoundSetter
	public void setAlmRestUserid(String almRestUserid) {
		this.almRestUserid = almRestUserid;
	}

	/**
	 * @return the almStopOnError
	 */
	public String getAlmStopOnError() {
		return almStopOnError;
	}
	/**
	 * @param almStopOnError the almStopOnError to set
	 */
	@DataBoundSetter
	public void setAlmStopOnError(String almStopOnError) {
		this.almStopOnError = almStopOnError;
	}

	/**
	 * @return the approveLevelRequest
	 */
	public String getApproveLevelRequest() {
		return approveLevelRequest;
	}
	/**
	 * @param approveLevelRequest the approveLevelRequest to set
	 */
	@DataBoundSetter
	public void setApproveLevelRequest(String approveLevelRequest) {
		this.approveLevelRequest = approveLevelRequest;
	}

	/**
	 * @return the approveReason
	 */
	public String getApproveReason() {
		return approveReason;
	}
	/**
	 * @param approveReason the approveReason to set
	 */
	@DataBoundSetter
	public void setApproveReason(String approveReason) {
		this.approveReason = approveReason;
	}

	/**
	 * @return the builderName
	 */
	public String getBuilderName() {
		return builderName;
	}
	/**
	 * @param builderName the builderName to set
	 */
	@DataBoundSetter
	public void setBuilderName(String builderName) {
		this.builderName = builderName;
	}

	/**
	 * @return the levelRequestAccess
	 */
	public String getLevelRequestAccess() {
		return levelRequestAccess;
	}
	/**
	 * @param levelRequestAccess the levelRequestAccess to set
	 */
	@DataBoundSetter
	public void setLevelRequestAccess(String levelRequestAccess) {
		this.levelRequestAccess = levelRequestAccess;
	}


}


src/main/java/com/ikanalm/plugins/jenkins/almclientlevelrequest/AlmclientLevelRequestStep.java
/**
 * 
 */
package com.ikanalm.plugins.jenkins.almclientlevelrequest;


import org.kohsuke.stapler.DataBoundConstructor;
import org.kohsuke.stapler.DataBoundSetter;
import org.kohsuke.stapler.QueryParameter;

import com.ikanalm.plugins.base.steps.BaseAntStep;

import hudson.Extension;
import hudson.Util;
import hudson.tasks.Ant;
import hudson.tasks.Ant.AntInstallation;
import hudson.util.FormValidation;
import hudson.util.ListBoxModel;
import jenkins.model.Jenkins;

/**
 * Class that defines the ALM ALM Client Level Request Create/Update/Control Phase as a Jenkins pipeline step.
 * Example usage :
 * almclientLevelRequest (antInstallation: "${antInstallation}",
 *      alm.baseUrl: "http://IKAN000:8080/alm",
 *      alm.buildNumber: "${buildNumber}",
 *      alm.deploysToExecute: "${deployEnvironments}",
 *      alm.description: "Build created by ${builderName} with Rest API",
 *      alm.levelRequest.levelName: "${levelName}",
 *      alm.levelRequest.oid: "",
 *      alm.package.name: "${packageName}",
 *      alm.project.name: "${projectName}",
 *      alm.projectStream.buildPrefix: "${buildPrefix}",
 *      alm.projectStream.buildSuffix: "${buildSuffix}",
 *      alm.redeliver: "${redeliver}",
 *      alm.rest.password: "",
 *      alm.rest.userid: "",
 *      alm.startDate: "",
 *      alm.status: "",
 *      alm.stopOnError: "true",
 *      alm.vcrTag: "",
 *      builderAction: "Create",
 *      builderName: "Client",
 *      levelRequest_access: "lroid",
 *      paramfile: "parameters.properties",
 *      status.maxwait_minutes: "30",
 *      status.success.mandatory: "true",
 *      status.waiting_seconds: "30"
 * )
 * 
 * Note that this Step CANNOT be used in a Jenkins Freestyle Project !
 *
 * @author IKAN
 *
 */
public class AlmclientLevelRequestStep extends BaseAntStep {

	private static final String MAIN_SCRIPT = "almclientLevelRequest.xml";
	private static final String FQ_PHASE_NAME = "com.ikanalm.plugins.jenkins.almclientlevelrequest";

    private String antInstallation;
    private String almBaseUrl;
    private String almBuildNumber;
    private String almDeploysToExecute;
    private String almDescription;
    private String almLevelRequestLevelName;
    private String almLevelRequestOid;
    private String almPackageName;
    private String almProjectName;
    private String almProjectStreamBuildPrefix;
    private String almProjectStreamBuildSuffix;
    private String almRedeliver;
    private String almRestPassword;
    private String almRestUserid;
    private String almStartDate;
    private String almStatus;
    private String almStopOnError;
    private String almVcrTag;
    private String builderAction;
    private String builderName;
    private String levelRequestAccess;
    private String paramfile;
    private String statusMaxwaitMinutes;
    private String statusSuccessMandatory;
    private String statusWaitingSeconds;

    @DataBoundConstructor
	public AlmclientLevelRequestStep(String antInstallation, String almBaseUrl, String almProjectName) {
	        this.antInstallation = antInstallation;
                this.almBaseUrl = almBaseUrl;
                this.almProjectName = almProjectName;

	}
        
	@Extension
	public static class DescriptorImpl extends BaseDescriptorImpl {
        /**
         * Display name of the Step in the Pipeline Syntax Snippet Generator
         */
		@Override 
		public String getDisplayName() {
            return "ALM Client Level Request Create/Update/Control";
        }
		/**
		 * Name of the function that is used to call the Step in a Pipeline script 
		 */
		@Override
		public String getFunctionName() {
			return "almclientLevelRequest";
		}
        /**
         * method that populates the configuration select named "antInstallation" with Ant installations
         * 
         * @return the antInstallation items
         */
        public ListBoxModel doFillAntInstallationItems() {
            ListBoxModel items = new ListBoxModel();
        	AntInstallation[] antInstallations = Jenkins.getInstanceOrNull().getDescriptorByType(Ant.DescriptorImpl.class).getInstallations();
            
        	for (AntInstallation antInstallation : antInstallations) {
				items.add(antInstallation.getName());
			}
        	
            return items;
        }

        /**
         * Field Validation
         * @param value the antInstallation
         * @return the antInstallation description
         */
		public FormValidation doCheckAntInstallation(@QueryParameter String value) {
			if (Util.fixEmptyAndTrim(value) == null) {
				return FormValidation.warning("Choose Ant Installation");
			}
			return FormValidation.ok();
		}

		public FormValidation doCheckAlmBaseUrl(@QueryParameter String value) {
			if (Util.fixEmptyAndTrim(value) == null) {
				return FormValidation.warning("IKAN ALM URL Login cannot be empty");
			}
			return FormValidation.ok();
		}

		public FormValidation doCheckAlmBuildNumber(@QueryParameter String value) {
			if (Util.fixEmptyAndTrim(value) == null) return FormValidation.ok();
			try {
				Integer.parseInt(value);
				return FormValidation.ok();
			} catch (NumberFormatException e) {
				return FormValidation.error("Specific ALM buildNumber to use for a deploy Level not numeric");
			}
		}

		public FormValidation doCheckAlmLevelRequestOid(@QueryParameter String value) {
			if (Util.fixEmptyAndTrim(value) == null) return FormValidation.ok();
			try {
				Integer.parseInt(value);
				return FormValidation.ok();
			} catch (NumberFormatException e) {
				return FormValidation.error("LR Oid parameter used when levelRequest_access=lroid not numeric");
			}
		}

		public FormValidation doCheckAlmProjectName(@QueryParameter String value) {
			if (Util.fixEmptyAndTrim(value) == null) {
				return FormValidation.warning("ALM Project Name where the Level Request is created or updated. cannot be empty");
			}
			return FormValidation.ok();
		}

	}

	/* (non-Javadoc)
	 * @see com.ikanalm.plugins.jenkins.base.BaseAntStep#getFQStepName()
	 */
	@Override
	public String getFQStepName() {
		return FQ_PHASE_NAME;
	}

	/* (non-Javadoc)
	 * @see com.ikanalm.plugins.jenkins.base.BaseAntStep#getMainScript()
	 */
	@Override
	public String getMainScript() {
		return MAIN_SCRIPT;
	}
	/**
	 * @return the antInstallation
	 */
	public String getAntInstallation() {
		return antInstallation;
	}

	/**
	 * @param antInstallation the antInstallation to set
	 */
	public void setAntInstallation(String antInstallation) {
		this.antInstallation = antInstallation;
	}

	/**
	 * @return the almBaseUrl
	 */
	public String getAlmBaseUrl() {
		return almBaseUrl;
	}
	/**
	 * @param almBaseUrl the almBaseUrl to set
	 */
	@DataBoundSetter
	public void setAlmBaseUrl(String almBaseUrl) {
		this.almBaseUrl = almBaseUrl;
	}

	/**
	 * @return the almBuildNumber
	 */
	public String getAlmBuildNumber() {
		return almBuildNumber;
	}
	/**
	 * @param almBuildNumber the almBuildNumber to set
	 */
	@DataBoundSetter
	public void setAlmBuildNumber(String almBuildNumber) {
		this.almBuildNumber = almBuildNumber;
	}

	/**
	 * @return the almDeploysToExecute
	 */
	public String getAlmDeploysToExecute() {
		return almDeploysToExecute;
	}
	/**
	 * @param almDeploysToExecute the almDeploysToExecute to set
	 */
	@DataBoundSetter
	public void setAlmDeploysToExecute(String almDeploysToExecute) {
		this.almDeploysToExecute = almDeploysToExecute;
	}

	/**
	 * @return the almDescription
	 */
	public String getAlmDescription() {
		return almDescription;
	}
	/**
	 * @param almDescription the almDescription to set
	 */
	@DataBoundSetter
	public void setAlmDescription(String almDescription) {
		this.almDescription = almDescription;
	}

	/**
	 * @return the almLevelRequestLevelName
	 */
	public String getAlmLevelRequestLevelName() {
		return almLevelRequestLevelName;
	}
	/**
	 * @param almLevelRequestLevelName the almLevelRequestLevelName to set
	 */
	@DataBoundSetter
	public void setAlmLevelRequestLevelName(String almLevelRequestLevelName) {
		this.almLevelRequestLevelName = almLevelRequestLevelName;
	}

	/**
	 * @return the almLevelRequestOid
	 */
	public String getAlmLevelRequestOid() {
		return almLevelRequestOid;
	}
	/**
	 * @param almLevelRequestOid the almLevelRequestOid to set
	 */
	@DataBoundSetter
	public void setAlmLevelRequestOid(String almLevelRequestOid) {
		this.almLevelRequestOid = almLevelRequestOid;
	}

	/**
	 * @return the almPackageName
	 */
	public String getAlmPackageName() {
		return almPackageName;
	}
	/**
	 * @param almPackageName the almPackageName to set
	 */
	@DataBoundSetter
	public void setAlmPackageName(String almPackageName) {
		this.almPackageName = almPackageName;
	}

	/**
	 * @return the almProjectName
	 */
	public String getAlmProjectName() {
		return almProjectName;
	}
	/**
	 * @param almProjectName the almProjectName to set
	 */
	@DataBoundSetter
	public void setAlmProjectName(String almProjectName) {
		this.almProjectName = almProjectName;
	}

	/**
	 * @return the almProjectStreamBuildPrefix
	 */
	public String getAlmProjectStreamBuildPrefix() {
		return almProjectStreamBuildPrefix;
	}
	/**
	 * @param almProjectStreamBuildPrefix the almProjectStreamBuildPrefix to set
	 */
	@DataBoundSetter
	public void setAlmProjectStreamBuildPrefix(String almProjectStreamBuildPrefix) {
		this.almProjectStreamBuildPrefix = almProjectStreamBuildPrefix;
	}

	/**
	 * @return the almProjectStreamBuildSuffix
	 */
	public String getAlmProjectStreamBuildSuffix() {
		return almProjectStreamBuildSuffix;
	}
	/**
	 * @param almProjectStreamBuildSuffix the almProjectStreamBuildSuffix to set
	 */
	@DataBoundSetter
	public void setAlmProjectStreamBuildSuffix(String almProjectStreamBuildSuffix) {
		this.almProjectStreamBuildSuffix = almProjectStreamBuildSuffix;
	}

	/**
	 * @return the almRedeliver
	 */
	public String getAlmRedeliver() {
		return almRedeliver;
	}
	/**
	 * @param almRedeliver the almRedeliver to set
	 */
	@DataBoundSetter
	public void setAlmRedeliver(String almRedeliver) {
		this.almRedeliver = almRedeliver;
	}

	/**
	 * @return the almRestPassword
	 */
	public String getAlmRestPassword() {
		return almRestPassword;
	}
	/**
	 * @param almRestPassword the almRestPassword to set
	 */
	@DataBoundSetter
	public void setAlmRestPassword(String almRestPassword) {
		this.almRestPassword = almRestPassword;
	}

	/**
	 * @return the almRestUserid
	 */
	public String getAlmRestUserid() {
		return almRestUserid;
	}
	/**
	 * @param almRestUserid the almRestUserid to set
	 */
	@DataBoundSetter
	public void setAlmRestUserid(String almRestUserid) {
		this.almRestUserid = almRestUserid;
	}

	/**
	 * @return the almStartDate
	 */
	public String getAlmStartDate() {
		return almStartDate;
	}
	/**
	 * @param almStartDate the almStartDate to set
	 */
	@DataBoundSetter
	public void setAlmStartDate(String almStartDate) {
		this.almStartDate = almStartDate;
	}

	/**
	 * @return the almStatus
	 */
	public String getAlmStatus() {
		return almStatus;
	}
	/**
	 * @param almStatus the almStatus to set
	 */
	@DataBoundSetter
	public void setAlmStatus(String almStatus) {
		this.almStatus = almStatus;
	}

	/**
	 * @return the almStopOnError
	 */
	public String getAlmStopOnError() {
		return almStopOnError;
	}
	/**
	 * @param almStopOnError the almStopOnError to set
	 */
	@DataBoundSetter
	public void setAlmStopOnError(String almStopOnError) {
		this.almStopOnError = almStopOnError;
	}

	/**
	 * @return the almVcrTag
	 */
	public String getAlmVcrTag() {
		return almVcrTag;
	}
	/**
	 * @param almVcrTag the almVcrTag to set
	 */
	@DataBoundSetter
	public void setAlmVcrTag(String almVcrTag) {
		this.almVcrTag = almVcrTag;
	}

	/**
	 * @return the builderAction
	 */
	public String getBuilderAction() {
		return builderAction;
	}
	/**
	 * @param builderAction the builderAction to set
	 */
	@DataBoundSetter
	public void setBuilderAction(String builderAction) {
		this.builderAction = builderAction;
	}

	/**
	 * @return the builderName
	 */
	public String getBuilderName() {
		return builderName;
	}
	/**
	 * @param builderName the builderName to set
	 */
	@DataBoundSetter
	public void setBuilderName(String builderName) {
		this.builderName = builderName;
	}

	/**
	 * @return the levelRequestAccess
	 */
	public String getLevelRequestAccess() {
		return levelRequestAccess;
	}
	/**
	 * @param levelRequestAccess the levelRequestAccess to set
	 */
	@DataBoundSetter
	public void setLevelRequestAccess(String levelRequestAccess) {
		this.levelRequestAccess = levelRequestAccess;
	}

	/**
	 * @return the paramfile
	 */
	public String getParamfile() {
		return paramfile;
	}
	/**
	 * @param paramfile the paramfile to set
	 */
	@DataBoundSetter
	public void setParamfile(String paramfile) {
		this.paramfile = paramfile;
	}

	/**
	 * @return the statusMaxwaitMinutes
	 */
	public String getStatusMaxwaitMinutes() {
		return statusMaxwaitMinutes;
	}
	/**
	 * @param statusMaxwaitMinutes the statusMaxwaitMinutes to set
	 */
	@DataBoundSetter
	public void setStatusMaxwaitMinutes(String statusMaxwaitMinutes) {
		this.statusMaxwaitMinutes = statusMaxwaitMinutes;
	}

	/**
	 * @return the statusSuccessMandatory
	 */
	public String getStatusSuccessMandatory() {
		return statusSuccessMandatory;
	}
	/**
	 * @param statusSuccessMandatory the statusSuccessMandatory to set
	 */
	@DataBoundSetter
	public void setStatusSuccessMandatory(String statusSuccessMandatory) {
		this.statusSuccessMandatory = statusSuccessMandatory;
	}

	/**
	 * @return the statusWaitingSeconds
	 */
	public String getStatusWaitingSeconds() {
		return statusWaitingSeconds;
	}
	/**
	 * @param statusWaitingSeconds the statusWaitingSeconds to set
	 */
	@DataBoundSetter
	public void setStatusWaitingSeconds(String statusWaitingSeconds) {
		this.statusWaitingSeconds = statusWaitingSeconds;
	}


}
